BSP/Src/BSP_I2C.c

```diff
@@ -0,0 +1,186 @@
+
+static QueueHandle_t I2C_MetaRegisterQueue;
+
+static QueueHandle_t I2C_ReceiveQueue; 
+
+HAL_StatusTypeDef BSP_I2C_Init() {
+    //QUEUEs
+    I2C_DataQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, __DataBufferStorage, &DataStaticQueue);
+    I2C_DataRegisterQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, __DataBufferStorage, &DataStaticQueue);
+    I2C_MetaQueue = xQueueCreateStatic(I2C_QUEUE_SIZE * 2, I2C_ITEM_LENGTH, __MetaBufferStorage, &MetaStaticQueue);
+    I2C_MetaRegisterQueue = xQueueCreateStatic(I2C_QUEUE_SIZE * 2, I2C_ITEM_LENGTH, __MetaBufferStorage, &MetaStaticQueue);
+    I2C_ReceiveQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, ReceiveBufferStorage, &ReceiveStaticQueue);
+
+    GPIO_InitTypeDef gpio_struct;
+    I2C_HandleTypeDef i2c_struct;
+
+    if (__HAL_RCC_I2C3_IS_CLK_DISABLED()) {
+        __HAL_RCC_I2C3_CLK_ENABLE();
+    }
+    if (__HAL_RCC_GPIOA_IS_CLK_DISABLED()) {
+        __HAL_RCC_GPIOA_CLK_ENABLE();
+    }
+    if (__HAL_RCC_GPIOC_IS_CLK_DISABLED()) {
+        __HAL_RCC_GPIOC_CLK_ENABLE();
+    }
+
+    uint32_t I2C_CLOCK_SPEED = 100000;
+    // i2c config 
+    i2c_struct.Init.ClockSpeed = I2C_CLOCK_SPEED;
+    i2c_struct.Init.DutyCycle = I2C_DUTYCYCLE_2;
+    // 7 bit addr mode
+    i2c_struct.Init.OwnAddress1 = 0x50 << 1;
+    i2c_struct.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+    i2c_struct.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    i2c_struct.Init.GeneralCallMode = I2C_GENERALCALL_ENABLED; 
+    i2c_struct.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+    HAL_StatusTypeDef stat = HAL_I2C_Init(&i2c_struct);
+
+    i2c_struct.Instance = I2C1;
+    HAL_NVIC_SetPriority(I2C1_EV_IRQn, 0, 0);
+    HAL_NVIC_EnableIRQ(I2C1_EV_IRQn);
+
+    return stat;
+}
+
+/** 
+ * @brief    preform I2C write
+ * @param    hi2c               struct with config data
+ * @param    deviceAddress      target device address
+ * @param    pDataBuff          data buffer
+ * @param    len                amount of data   
+ * @return   void
+ */
+bool BSP_I2C_Write(I2C_HandleTypeDef *hi2c, 
+              uint8_t deviceAddress, 
+              uint8_t* pDataBuff, 
+              uint16_t len) {
+    if (!__HAL_I2C_GET_FLAG(hi2c, I2C_FLAG_BUSY)) {
+        // HAL requires 7b I2C address left shifted
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAddress << 1, *pDataBuff, len);
+    }
+    else if (uxQueueSpacesAvailable(I2C_DataQueue) >= len) {
+            
+            // HAL requires 7b I2C address left shifted
+            struct metaInfo metaData = {.deviceAddr = deviceAddress << 1, .length = len};
+            xQueueSend(I2C_MetaQueue, (void *) &metaData, 0);
+    }
+}
+
+/**
+ * @brief Recursively go through to see if there is any more info that needs to be transmitted,
+ * if there is transmit it
+ * @param hi2c                  I2C handle that interrupted (passed in by IRQ)
+ */
+void HAL_I2C_MasterTxCpltCallback(I2C_HandleTypeDef *hi2c) {
```

Comment:
support multiple i2cs maybe?

Reply:
im 99% sure the hi2c is passed in by the interrupt handler behind the scenes, its not something we call directly, so the i2c1/i2c3 objects will get passed in magically when their interrupts trigger